File: go/dijkstra_algorithm.go
package main

import (
    "container/heap"
    "fmt"
    "math"
)


type Edge struct {
    to     int
    weight int
}


type PriorityQueueItem struct {
    node     int
    distance int
}


type PriorityQueue []PriorityQueueItem

func (pq PriorityQueue) Len() int { return len(pq) }
func (pq PriorityQueue) Less(i, j int) bool {
    return pq[i].distance < pq[j].distance
}
func (pq PriorityQueue) Swap(i, j int) { pq[i], pq[j] = pq[j], pq[i] }

func (pq *PriorityQueue) Push(x interface{}) {
    *pq = append(*pq, x.(PriorityQueueItem))
}

func (pq *PriorityQueue) Pop() interface{} {
    old := *pq
    n := len(old)
    item := old[n-1]
    *pq = old[0 : n-1]
    return item
}

// Dijkstra’s Algorithm
func dijkstra(graph [][]Edge, start int) []int {
    n := len(graph)
    dist := make([]int, n)
    for i := 0; i < n; i++ {
        dist[i] = math.MaxInt32 // infinity
    }
    dist[start] = 0

    pq := &PriorityQueue{}
    heap.Init(pq)
    heap.Push(pq, PriorityQueueItem{node: start, distance: 0})

    for pq.Len() > 0 {
        cur := heap.Pop(pq).(PriorityQueueItem)
        u := cur.node
        d := cur.distance

        // Skip if we already found a better path
        if d > dist[u] {
            continue
        }

        // Relax edges
        for _, edge := range graph[u] {
            v := edge.to
            w := edge.weight
            if dist[u]+w < dist[v] {
                dist[v] = dist[u] + w
                heap.Push(pq, PriorityQueueItem{node: v, distance: dist[v]})
            }
        }
    }

    return dist
}

func main() {
    /*
       Example graph (undirected):
       0 --4--> 1
       0 --1--> 2
       2 --2--> 1
       1 --1--> 3
       2 --5--> 3

       Represented as adjacency list:
    */
    graph := [][]Edge{
        {{1, 4}, {2, 1}},    // edges from 0
        {{0, 4}, {2, 2}, {3, 1}}, // edges from 1
        {{0, 1}, {1, 2}, {3, 5}}, // edges from 2
        {{1, 1}, {2, 5}},    // edges from 3
    }

    start := 0
    dist := dijkstra(graph, start)

    fmt.Printf("Shortest distances from node %d:\n", start)
    for i, d := range dist {
        fmt.Printf("To %d = %d\n", i, d)
    }
}

File: rust/dijkstra_algorithm.rs
use std::collections::BinaryHeap;
use std::cmp::Ordering;

#[derive(Copy, Clone, Eq, PartialEq)]
struct State {
    cost: usize,
    position: usize,
}

// Implement `Ord` for min-heap using BinaryHeap
impl Ord for State {
    fn cmp(&self, other: &Self) -> Ordering {
        // Notice we flip the ordering here
        other.cost.cmp(&self.cost)
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

// Edge structure
#[derive(Copy, Clone)]
struct Edge {
    node: usize,
    weight: usize,
}

// Dijkstra’s Algorithm
fn dijkstra(graph: &Vec<Vec<Edge>>, start: usize) -> Vec<usize> {
    let n = graph.len();
    let mut dist = vec![usize::MAX; n];
    let mut heap = BinaryHeap::new();

    dist[start] = 0;
    heap.push(State { cost: 0, position: start });

    while let Some(State { cost, position }) = heap.pop() {
        if cost > dist[position] {
            continue;
        }

        for edge in &graph[position] {
            let next = State { cost: cost + edge.weight, position: edge.node };
            if next.cost < dist[next.position] {
                dist[next.position] = next.cost;
                heap.push(next);
            }
        }
    }

    dist
}

fn main() {
    // Graph represented as adjacency list
    // Each node has a vector of edges (neighbor, weight)
    let graph = vec![
        vec![Edge { node: 1, weight: 4 }, Edge { node: 2, weight: 1 }],
        vec![Edge { node: 3, weight: 1 }],
        vec![Edge { node: 1, weight: 2 }, Edge { node: 3, weight: 5 }],
        vec![],
    ];

    let start = 0;
    let distances = dijkstra(&graph, start);

    for (i, &d) in distances.iter().enumerate() {
        if d == usize::MAX {
            println!("Vertex {} is unreachable", i);
        } else {
            println!("Distance from {} to {} is {}", start, i, d);
        }
    }
}
